New User Setup and Input Overlay

JamesVanBoxtel opened the pull request:
- Language selection setup screen created, shown when language not set
- Discord information setup screen created, new config for when it has been shown
- New Input Overlay Scene for picking inputs
- New controller theme graphics for representing input device type
- Comprehensive BattleRoom methods for managing players and their input assignments
- Restore previous assignments in 1P if they have been picked this session
- Language now defaults to English but not confirmed until picked
- Set Language method decoupled from saving language to support above
- InputPromptRenderer for drawing device images and eventually button prompts
- InputConfiguration has become a full class with device name, image, label etc fully accessible.
- InputManager manages the input configurations and provides a touch one.
- Added a second WASD default configuration for when new players don't have any yet and expect to navigate menus with WASD
- Input Device Overlay is shown in character select when not all human local players are assigned.
- Added change input device button for changing after you have already assigned
- Input Config Menu visuals improved, added auto configure joystick
- Added SceneCoordinator class for auto showing next scenes based on state
- Added ability for labels to autosize
- Added ability for menus to size to fit
- Added a slider menu item
- Made UIElement properly traverse touch from topmost down
- Auto show input config on new controller
- Add way to set TYPE variable on class constructor
- Added debugging print functions on UIElement
- Clear previous bindings when reusing key
- Require all keys bound before exiting config
- Input manager cleanup

Video showing language picker, discord info, configuring new controller, can't exit till not partially configured.


https://github.com/user-attachments/assets/7fd75dca-5a57-4374-a4c5-a2a4f5b46aa1

JamesVanBoxtel committed:
New User Setup and Input Overlay

Language selection setup screen created, shown when language not set
Discord information setup screen created, new config for when it has been shown
New Input Overlay Scene for picking inputs
New controller theme graphics for representing input device type
Comprehensive BattleRoom methods for managing players and their input assignments
Restore previous assignments in 1P if they have been picked this session
Language now defaults to English but not confirmed until picked
Set Language method decoupled from saving language to support above
InputPromptRenderer for drawing device images and eventually button prompts
InputConfiguration has become a full class with device name, image, label etc fully accessible.
InputManager manages the input configurations and provides a touch one.
Added a second WASD default configuration for when new players don't have any yet and expect to navigate menus with WASD
Input Device Overlay is shown in character select when not all human local players are assigned.
Added change input device button for changing after you have already assigned
Input Config Menu visuals improved, added auto configure joystick
Added SceneCoordinator class for auto showing next scenes based on state
Added ability for labels to autosize
Added ability for menus to size to fit
Added a slider menu item
Made UIElement properly traverse touch from topmost down
Auto show input config on new controller
Add way to set TYPE variable on class constructor
Added debugging print functions on UIElement
Clear previous bindings when reusing key
Require all keys bound before exiting config
Input manager cleanup

JamesVanBoxtel committed:
Add Debug Settings Architecture

Repost of this commit since it was reverted.

The Debug Settings system provides runtime-configurable debug features through a UI overlay. Settings are persisted to config.debug and can be accessed anywhere via the DebugSettings singleton.

Also provide a way to generate the debug menu and surface it from a debug button.

To support this new button and future overlays, add a root element to the game so we can order overlays

Updated Bool Selector to size correctly
Updated Style Selector creation to make sure sizing and layout is right
Reload main menu when you come back to it
Make navigation stack a UIElement so it works with touch handling

Fixes #697

JamesVanBoxtel committed:
Merge branch 'beta' into initialSetupAndInputPickerFinal

JamesVanBoxtel committed:
Merge branch 'beta' into initialSetupAndInputPickerFinal

Endaris commented:
Hi, some initial feedback, I hope to make some more serious progress over the weekend but there's something I already want to discuss regarding the architecture adjustment to the scenes. I wanted to review some yesterday and then got stuck superhard on this really early which kept me from looking at the rest so I'll get this one out early

In particular this is about `SceneCoordinator` and the new `Scene.triggerNextScene` callback.

Obviously it works and it's good in the sense that the scenes themselves don't have to know about how they are used but I think the approach strays a bit too far from what a Scene ideally should be to keep things easily composable without having to add a class like SceneCoordinator to glue things together.

Semantically I understand a Scene as the composition of audio, UI and input options that serves a specific purpose. For the feature in this PR I would intuitively think that what is currently `StartUp` could be renamed to `BootScene` and that what the SceneCoordinator does is more representative of what a `StartUpScene` does - except that it is not a Scene right now.

I think having a `SceneCoordinator` besides the `NavigationStack` is confusing and misleading.
Partially a name thing but generally speaking, the first start up functionality could mostly also be implemented by checking in the current `StartUp` whether the respective config fields are set and simply pushing the desired scenes in reverse order on top of the NavigationStack and then they just pop off one after another and always set the config field on exit. And the callback for bringing up the InputConfigMenu could live on `Game` rather than have its own class.
I think if we feel like there are lacking options in regards to chaining scenes, then we should extend the functionality of NavigationStack rather than inventing new things that intentionally work around the center component of our Scene navigation. This is the main gripe I have with this new class.

I think we have two options here:

1. Aim for composition of menus instead of composition of scenes.
Menus can be ready made UIElements that provide an onExit callback so that a scene can chain multiple menus after another by setting the callback on the menu instance and replacing its own uiRoot with the new menu. We could gather standalone menus in a new directory, e.g. `src/menus`.
If we have menus we want to show in different situations like the InputConfigMenu, we have the option to wrap them in a standalone scene but also present them in a composed Scene like the first start up assistant here.
SceneCoordinator becomes an actual StartUpScene. It uses the exit callback of the new menus to call its own `continueSetupFlow` which replaces the menu within the scene itself instead of pushing a Scene.
Flexible, does not collide with the NavigationStack purpose. Seems like an intuitive architecture to break up OptionsMenu down the line.

3. Add more functions to NavigationStack that help with pushing a sequence of scenes.
Could look something like this in place of where `SceneCoordinator.handleStartupComplete(SceneCoordinator)` is called in `StartUp`?
```Lua
local startUpSequence = {}
-- Check language selection
if not config.language_code then
  table.insert(startUpSequence, require("client.src.scenes.LanguageSelectSetup"))
end

-- Check Discord community welcome
if not config.discordCommunityShown then
  table.insert(startUpSequence, require("client.src.scenes.DiscordCommunitySetup"))
end

-- Check for unconfigured joysticks
if input.hasUnsavedChanges or input:hasUnconfiguredJoysticks() then
  table.insert(startUpSequence, require("client.src.scenes.InputConfigMenu"))
end

table.insert(startUpSequence, require("client.src.scenes.MainMenu"))

GAME.navigationStack:pushSequence(startUpSequence)
```
The new `pushSequence` would respectively take care of pushing the scenes in the correct order so they pop up one after another.


I think I'm fine with either, I think you were leaning towards menus as UI elements before?
But it's mainly food for thought based on my feeling of SceneCoordinator adding complexity to the project when there's less complex or more flexible solutions.

JamesVanBoxtel commented:
I think I see what you mean, here is a proposal for a cleaner approach. Please let me know if its what you were thinking.

  Instead of having a separate SceneCoordinator class that works alongside the navigation system, I'll rename the BootScene to StartupScene and move all the SceneCoordinator logic into it.

  1. Uses NavigationStack directly as the sole navigation mechanism—no separate coordinator layer, no Scene.triggerNextScene callback
  2. Owns the startup flow by checking what setup steps are needed and pushing the appropriate scenes (LanguageSelect, DiscordCommunity, InputConfig) via GAME.navigationStack:push()

  The flow would work like this: StartUpScene enters and evaluates what's needed, pushes the first required setup scene, then when that scene pops, StartUpScene:resume() re-evaluates and either
  pushes the next step or replaces itself with MainMenu when done. All navigation goes through NavigationStack—the startup scene just decides what to navigate to. This aligns with your suggestion
  of working within NavigationStack rather than inventing parallel mechanisms.

  I may need to add a little bit of logic to NavigationStack so we can transition to the next scene without showing the boot scene for a frame and getting a flicker. TBD

  For the input config callback you mentioned should live on Game, I'll move that there or just call GAME.navigationStack:push(InputConfigMenu) directly from the startup scene.

  This eliminates having a coordinator side mechanism and generic callback that were the core concerns, while still giving us a proper place to orchestrate the first-time setup flow.

  Let me know if this works for you and I can implement.

JamesVanBoxtel commented:
After attempting to implement the solution we discussed, I encountered a fundamental issue that I believe I encountered during my initial attempt.

We require a seamless transition from Discord to InputConfig, with no flickering animation. This becomes even more challenging when we want to implement a slide transition or something similar, as both scenes are simultaneously visible. We cannot have an “in-between” scene that is not part of the actual transition. Instead, we need to replace one scene with another without briefly displaying something in between.

BattleRoom, ChallengeMode, and Game all already coordinate scenes that perform similar functions.

Additionally, we have the following requirements:

1. Scenes must be self-contained. DiscordCommunitySetup should not be aware of InputConfig.
2. There must be a mechanism to determine the flow of the transition. The logic for “After Discord, show InputConfig” should reside somewhere.

This is why I implemented the solution in the manner I did and added Scene.triggerNextScene to decouple the components. However, I believe triggerNextScene should not be generic, and each scene should have its own specific requirements for the name or if a function is required for returning. For instance, ChallengeMode could have a win and a lose function passed in.

Therefore, I believe the original solution is largely sound, except for the fact that triggerNextScene should not be generic. Please advise whether you believe modifying that is sufficient or if we still need to make further changes.

If so, how can we achieve these goals in an alternative manner? Alternatively, would you like to attempt refactoring while still meeting these goals?

Endaris committed:
rename StartUp to BootScene

Endaris committed:
replace triggerNextScene pattern with a simple conditional push sequence within the BootScene

Endaris committed:
move code for triggering InputConfigMenu on top into Game and remove SceneCoordinator

Endaris commented on client/src/ui/UIElement.lua:
```diff
@@ -259,4 +269,41 @@ function UIElement:handleFocusedInput(inputs, dt)
+  end
+
+  return table.concat(lines, "\n")
+end
```
Rename these to something like `toStringWithDepth` and `toString`?

Endaris commented on common/engine/Match.lua:
```diff
@@ -12,6 +12,7 @@ local LegacyPanelSource = require("common.compatibility.LegacyPanelSource")
 local InputCompression = require("common.data.InputCompression")
 local ReplayV3 = require("common.data.ReplayV3")
 local MatchRules = require("common.data.MatchRules")
+local DebugSettings = require("client.src.debug.DebugSettings")
```
Obsolete leftover that is no longer needed after we put in 292f5fda46676a6f9a9913677b1b2e8b97f3e56f

Endaris commented on main.lua:
```diff
@@ -187,6 +187,18 @@ function love.joystickreleased(joystick, button)
   inputManager:joystickReleased(joystick, button)
 end
 
+-- Intentional override
+---@diagnostic disable-next-line: duplicate-set-field
+function love.joystickadded(joystick)
+  GAME:onJoystickAdded(joystick)
+end
+
+-- Intentional override
+---@diagnostic disable-next-line: duplicate-set-field
+function love.joystickremoved(joystick)
+  inputManager:onJoystickRemoved(joystick)
+end
```
A bit unintuitive that one is routed through GAME and the other only through inputManager.
I see why but it might be consistent to just proxy joystickremoved through GAME as well.

Endaris commented on client/src/config.lua:
```diff
@@ -130,12 +131,15 @@ config = {
       function()
+        local encoded = json.encode(config)
+        ---@cast encoded string
         love.filesystem.write("conf.json", json.encode(config))
```
Assume you wanted to pass `encoded` in here as an argument?

Endaris committed:
move unregister process for joysticks into joystickManager
remove return from updating the cache, just call the getter instead in the one location it is required

Endaris commented on common/lib/joystickManager.lua:
```diff
@@ -167,28 +170,4 @@ function love.joystickadded(joystick)
   joystickManager.devices[id] = device
 end
 
--- Intentional override
----@diagnostic disable-next-line: duplicate-set-field
-function love.joystickremoved(joystick)
-  -- GUID identifies the device type, 2 controllers of the same type will have a matching GUID
-  -- the GUID is consistent across sessions
-  local guid = joystick:getGUID()
-  -- ID is a per-session identifier for each controller regardless of type
-  local id = joystick:getID()
-
-  local vendorID, productID, productVersion = joystick:getDeviceInfo( )
-
-  logger.info("Disconnecting device " .. vendorID .. ";" .. productID .. ";" .. productVersion .. ";" .. joystick:getName() .. ";" .. guid .. ";" .. id)
-
-  if joystickManager.guidsToJoysticks[guid] then
-    joystickManager.guidsToJoysticks[guid][id] = nil
-
-    if tableUtils.length(joystickManager.guidsToJoysticks[guid]) == 0 then
-      joystickManager.guidsToJoysticks[guid] = nil
-    end
-  end
-
-  joystickManager.devices[id] = nil
-end
```
Good idea to not have the love callback here.
But logically this still should be a function on joystickManager just like the register function.

Endaris commented on common/lib/joystickManager.lua:
```diff
@@ -101,9 +101,12 @@ function joystickManager:getDPadState(joystick, hatIndex)
+
+  if joystickManager.devices[joystick:getID()] then
+    return
+  end
```
I think I prefer checking the inverse outside the function to make the code calling this more clear.
Otherwise, everytime I check `joystickPressed` I'll wonder "why is the joystick being registered on every single press?"

Endaris committed:
extract check for joystick being registered on inputs for more clarity

Endaris committed:
rename write_key_file to writeKeyConfigurationToFile

Endaris commented on client/src/inputManager.lua:
```diff
@@ -439,25 +464,75 @@ function inputManager:getSaveKeyMap()
   return result
 end
 
+function inputManager:write_key_file()
```
Rename to match our naming conventions if it's already being moved

Endaris committed:
rename all local config variables to inputConfig to prevent confusion with the config global

Endaris commented on client/src/inputManager.lua:
There are several uses of `config` as local variables / parameters.
This can be a bit confusing with shadowing our `config` global, so I would rename them to `inputConfig` just to be explicit.

Endaris committed:
add some annotations for inputManager/InputConfiguration

Endaris commented on client/src/inputManager.lua:
```diff
@@ -488,4 +563,331 @@ function inputManager:releaseConfiguration(player, inputConfiguration)
+  end
+
+  return not assignedConfigs[activeConfig]
+end
```
Given that this function does not directly depend on `battleRoom` and the only place it is called already guarantees there is a battleRoom, it should take the output of `battleRoom:getLocalHumanPlayers()` directly as a parameter.

Endaris committed:
change checkForUnassignedConfigurationInputs to directly take a player array as its argument

Endaris committed:
proxy joystickRemoved through Game and fix extra encode for config write

Endaris committed:
rename UiElement functions for generating debug strings

Endaris committed:
remove obsolete require from Match

Endaris reviewed (commented):
I did make 2 PRs on your branch on your repo.
One for the thing discussed previously.
The other is addressing the comments made for this review section so that it's a bit easier to parse individually why each change was made. I tried to make my changes commit by commit so you can easily revert some if you disagree.

I'm not quite done looking at everything but the remaining parts are mostly the new screens/UI elements which should be a bit faster as I'll have less preconceived ideas about them.

Endaris committed:
readd ModValidationScene (forgot about it)

Endaris commented on client/assets/localization.csv:
```diff
@@ -638,3 +638,28 @@ mod_manage_submods,Label for submods in mod management,Sub Mods,Sous-mods,Submod
+tauntup,Input configuration label for taunt up button,Taunt Up,Provocation Haut,Provocação Cima,挑発上,Burla Arriba,Spott Oben,Provocazione Su,เยาะเย้ยขึ้น
+tauntdown,Input configuration label for taunt down button,Taunt Down,Provocation Bas,Provocação Baixo,挑発下,Burla Abajo,Spott Unten,Provocazione Giù,เยาะเย้ยลง
+change_input_device,Label for changing input device,Change Input Device,Changer de périphérique d'entrée,Alterar dispositivo de entrada,入力デバイスを変更,Cambiar dispositivo de entrada,Eingabegerät ändern,Cambia dispositivo di input,เปลี่ยนอุปกรณ์ควบคุม
+press_button_device,Prompt to press a button on desired input device,Press a button on the device you want to use,Appuyez sur un bouton du périphérique que vous souhaitez utiliser,Pressione um botão no dispositivo que deseja usar,使用したいデバイスのボタンを押してください,Presiona un botón en el dispositivo que quieres usar,Drücke eine Taste auf dem Gerät\, das du verwenden möchtest,Premi un pulsante sul dispositivo  ...
```
I was a bit confused by this because I kept pressing a button and nothing happened when really it wanted me to hold one instead.

Endaris commented on client/assets/localization.csv:
```diff
@@ -638,3 +638,28 @@ mod_manage_submods,Label for submods in mod management,Sub Mods,Sous-mods,Submod
+swap1,Input configuration label for first swap button,Swap 1,Échanger 1,Trocar 1,スワップ1,Intercambiar 1,Tausch 1,Scambia 1,สลับ 1
+swap2,Input configuration label for second swap button,Swap 2,Échanger 2,Trocar 2,スワップ2,Intercambiar 2,Tausch 2,Scambia 2,สลับ 2
+raise1,Input configuration label for first raise button,Raise 1,Monter 1,Levantar 1,上げる1,Elevar 1,Heben 1,Alza 1,เลื่อน 1
+raise2,Input configuration label for second raise button,Raise 2,Monter 2,Levantar 2,上げる2,Elevar 2,Heben 2,Alza 2,เลื่อน 2
```
We may want to indicate the non-gameplay functions here as well to communicate that Swap1 has "select/confirm" functionality, Swap2 has "back" functionality and that raise1/2 may have turn page functionality on some menus.

Endaris commented on client/assets/localization.csv:
```diff
@@ -638,3 +638,28 @@ mod_manage_submods,Label for submods in mod management,Sub Mods,Sous-mods,Submod
 op_about_puzzles,,About custom puzzles,À propos des puzzles personnalisés,Sobre quebra-cabeças personalizados,カスタムパズルについて,Acerca de rompecabezas personalizados,How to: Eigene Puzzles erstellen,A proposito dei puzzle personalizzabili,เกี่ยวกับ custom puzzles
+discord_welcome_title,Title for Discord community welcome screen,Welcome to Panel Attack!,Bienvenue dans Panel Attack!,Bem-vindo ao Panel Attack!,パネルアタックへようこそ！,¡Bienvenido a Panel Attack!,Willkommen bei Panel Attack!,Benvenuti in Panel Attack!,ยินดีต้อนรับสู่ Panel Attack!
+discord_message_line1,First line of Discord welcome message,"Join our Discord to meet players, share ideas, and talk all things Panel Attack.","Rejoignez notre Discord pour rencontrer des joueurs, partager vos idées et discuter de tout ce qui concerne Panel Attack.","Junte-se ao nosso Discord para conhecer jogadores, compartilhar ideias e falar tudo sobre Panel Attack.","Discordに参加してプレイヤーと出会い、アイデ ...
+discord_message_line2,Second line of Discord welcome message,"Show off your mods and art, rediscover classic characters and stages, and create exciting new ones.","Présentez vos mods et vos créations, redécouvrez des personnages et des stages classiques, et créez-en de nouveaux passionnants.","Mostre seus mods e artes, redescubra personagens e fases clássicas e crie novidades empolgantes.","自分のMO ...
```
German localization
line2: "Zeige deine Mods und Kunst, entdecke bekannte Charaktere und Arenen neu und erschaffe deine eigenen."

Endaris commented on client/src/ui/ChangeInputButton.lua:
```diff
@@ -0,0 +1,207 @@
+    connectionInfo.player:disconnectSignal("inputConfigurationChanged", connectionInfo.connection)
+  end
+  self.signalConnections = {}
+end
```
This is only called by setBattleRoom which is unused itself.
Did you have a specific plan in mind for trying to track the signalConnections like this?
`connectSignal` does not return a connection object at the moment so I suspect these might be some leftovers from an approach that did not work as imagined.

Endaris committed:
Merge branch 'JamInputPickerSuggestions2' into JamInputPicker

Endaris committed:
update german localization and change wording for input device selection from "press" to "hold" as short presses don't select the device

Endaris committed:
add a disclaimer label to warn users about the quality of the translation and make them aware that contributions are possible

Endaris committed:
remove unused function from Player

Endaris committed:
make ChangeInputButton use Players array directly and remove unused code
fix config not saving

Endaris committed:
consolidate ChangeInputButton by making the inputdevice assignment clear part of its core functionality
relieve CharacterSelect/BattleRoom of some glue code that could live on Player/ChangeInputButton instead

Endaris committed:
open InputDeviceOverlay actively through CharacterSelect instead of having it spring up on its own

Endaris committed:
fix annotations for InputDeviceOverlay

Endaris committed:
make InputDeviceOverlay work directly with players instead of going through BattleRoom

Endaris committed:
remove unused functions from BattleRoom that have been moved to Player/Input components

Endaris committed:
patch up some annotations and remove fuzziness when local inputConfig is unclear during receipt of matchStart message

Endaris committed:
always set inputMethod with restrictInputs and don't reset it with unrestrict
that way it is more likely that the inputMethod matches for a lastUsedInputConfig recovery in case of a matchStart

Endaris committed:
fix construction args of InputDeviceOverlay for PuzzleMenu

Endaris committed:
update localization some more, widen labels for keybinds to fit the text
fixed PuzzleMenu not opening the inputoverlay

Endaris committed:
fix localization not being refreshed after selecting language in the languageselectsetup screen

Endaris committed:
always show the input menu on first start up, just to be sure
